### ISSUE TITLE
Types, hints, issues and approach

### ISSUE THREAD

**Issue by markph0204** [pygame/pygame#839]
@illume provided a good approach/plan I wanted to capture and mark off as we progress:

- [ ] document types for the html(sphinx) and IDE documentation reading
- [ ] allow types to be used for type checking tools(mypy, pylint, jedi, etc), and/or IDEs
- [ ] types to be used for completion in IDEs (and tools like jedi)
- [ ] somehow work with our python, C, Cython, and rst
- [ ] figure out how we can do it incrementally (eg, maybe just Surface and a few other modules to start with).
- [ ] hopefully getting something useful released in the near term for 'core' types and modules in (1.9.5 release perhaps).
- [ ] hopefully not too much duplication of types if possible (eg, specifying in .rst, .pyi, and .py files)
- [ ] hopefully don't increase load time too much

Also for reference [[types](https://github.com/pygame/pygame/issues?utf8=✓&q=is%3Aissue+is%3Aopen+label%3Atypes+)] marked issues.

**Comment by markph0204** [comment]
ref: https://www.python.org/dev/peps/pep-0561/

"For package maintainers wishing to ship stub files containing all of their type information, it is preferred that the *.pyi stubs are alongside the corresponding *.py files."

**Comment by markph0204** [comment]
PR: https://github.com/pygame/pygame/pull/871

Provides type hints in IDEA type editors (PyCharm)
![image](https://user-images.githubusercontent.com/210862/53681349-5c31e900-3cb6-11e9-8eee-2b3102f95684.png)

**Comment by NotaSmartDev** [comment]
Hey, I really like the idea but I'm not using PyCharm. Might do one day but as of now it is not the case. The IDE I'm using is SublimeText and I wondered if it'd be possible (and not too hard) for you guys to do something alike it ? I have no idea if SublimeText supports this kind of hints or not though.

**Comment by markph0204** [comment]
@NotaSmartDev Yes this will work with Sublime text as well. Just use the plugin called **SublimeJEDI**.

* Sublime Text (SublimeJEDI [ST2 + ST3], anaconda [only ST3])

ref: https://github.com/davidhalter/jedi

**Comment by NotaSmartDev** [comment]
Thanks you very much ! I've just installed Anaconda and it works fine if a little slow at first. Would you recommend me to use SublimeJEDI instead ? It's a really nice feature btw. Allows me not to look at the docs just as often.

**Comment by markph0204** [comment]
This feature has not been merged yet.  Did you build from the branch?  I only recommend that plugin because it is using Jedi.  I love the feedback -- thank you.  PR #871

**Comment by NotaSmartDev** [comment]
Yes I did, I just couldn't wait for it to be merged. Thank you for making this.

**Comment by markph0204** [comment]
Great! This helps with validation. More to come!

**Comment by illume** [comment]
Hi @markph0204,

Nice work testing these :)

**Comment by NotaSmartDev** [comment]
I wonder, if there is any plan to merge it soon though ?

**Comment by markph0204** [comment]
Yes it has been merged -- more improvements to come once I free up some
time.

**Comment by illume** [comment]
I updated the issue description to make the checklist into questions we've answered, and things that have been tested already.

There are some open questions left, but I think we are getting closer to having a clear idea of how we will do this.

**Comment by bendavis78** [comment]
Does code completion work yet with Jedi? I'm using vscode, and the following doesn't work:

    import pygame
    
    screen = pygame.display.set_

I expect  "set_mode" to come up, but it doesn't.

**Comment by SurealCereal** [comment]
It seems the PR #871 to add the .pyi files was commented out and left that way: https://github.com/pygame/pygame/blob/7719d927417be16e6f6c9dfdb81cce2530883d9d/setup.py#L335

**Comment by AmjadHD** [comment]
@s0lst1ce
>Would you recommend me to use SublimeJEDI instead ?
Use the [LSP](https://packagecontrol.io/packages/LSP) package with the [pyls](https://github.com/palantir/python-language-server)
Here are the docs https://lsp.readthedocs.io/en/latest/

**Comment by s0lst1ce** [comment]
Thanks I've found my way some time ago though. I appreciate the thought though :) @AmjadHD

**Comment by illume** [comment]
@MightyJosip added **lots** more type hints in here https://github.com/pygame/pygame/pull/1685

Also type hints will be installed with the next `pygame==2.0.0.dev8` release (unless something pops up in testing for why we shouldn't). There's some discussion in the PR about future improvements, and things to investigate.

**Comment by judej** [comment]
I have spent some time cleaning up the type stubs in the buildconfig folder and would like to contribute them. Would you like me to submit a PR for this?

**Comment by judej** [comment]
Alternatively, I can submit them to typeshed. I have tested the stubs using VSCode and Pylance Language server and I get good completions and sigs etc. I can most of the examples in strict and they are mostly clean - no diagnostics.

Would like to know which option is preferable. thanks!

**Comment by MyreMylar** [comment]
A PR sounds great to me. @MightyJosip has been doing a lot of work on these this year so will probably want to give it a looksie.

**Comment by MightyJosip** [comment]
Hi @judej. Thanks for your work and you can freely make PR for it. I have tested typehints with PyCharm and mypy, but VSCode is also good (a lot of people use it so it is important that type hints work in VSC). We discussed submitting typehints to typeshed, but for now the decision is to not do that since there are still a lot of changes in dev versions of pygame (once we officially release pygame2 than we will reconsider it again)

**Comment by McSinyx** [comment]
I'm new to the whole type annotation thing, but today I tried to run mypy on a project using pygame and got something like

    error: Skipping analyzing 'pygame': found module but no type hints or library stubs

I looked up [PEP 561](https://www.python.org/dev/peps/pep-0561/#packaging-type-information) and it states that

> Package maintainers who wish to support type checking of their code MUST add a marker file named `py.typed` to their package supporting typing.

So I `touch .../site-packages/pygame/py.typed` and mypy is now happy.  I wonder if the file wasn't included because the maintainers don't think the stubs are ready or that it was forgotten.

**Comment by MyreMylar** [comment]
Seems to be a thing only used by mypy from a quick google. [Here](https://github.com/pytorch/pytorch/pull/18511) is a PR where they added the blank file to pyTorch to support mypy, and there is a discussion [here](https://github.com/encode/httpx/issues/193) on how dumb this is, but they don't seem to have found an alternative. It seems to be an official PEP as of python 3.7.

It's literally just an empty file with the name `py.typed` that needs to be passed around with the installed package. My suggestion is that we could dynamically create it in setup.py to avoid having it in the repo?

**Comment by Zeelot** [comment]
hey folks, I'm running into some typing issues but I'm not sure if this is the right place to ask for help. I use `mypy` and it is complaining when I use `pygame.Surface` instead of `pygame.surface.Surface` as my types. Are these meant to be different types or is `pygame.Surface` supposed to be a shortcut to `pygame.surface.Surface`? I have the same issue with `pygame.Vector2` vs `pygame.math.Vector2`.

```
Incompatible types in assignment (expression has type "pygame.math.Vector2", variable has type "pygame.Vector2")
Incompatible return value type (got "pygame.surface.Surface", expected "pygame.Surface")
```

I think the issue is because `pygame.Surface` seems to be defined as a class that extends `pygame.surface.Surface` instead of just being defined as the same class. So things like `pygame.display.set_mode()` will always return instances of `pygame.surface.Surface` and not `pygame.Surface`.

Can anyone help me better understand what I might be doing wrong? Are these meant to be different classes or is this a bug?

**Comment by Zeelot** [comment]
Just to verify, I edited `pygame/__init__.pyi` and set the classes to simple variables and this seems to have resolved my issues, but I'm not comfortable enough with `pygame` to know if I broke any assumptions:

```python
# This classes are auto imported with pygame, so I put their declaration here
Rect = pygame.rect.Rect
Surface = pygame.surface.Surface
Color = pygame.color.Color
PixelArray = pygame.pixelarray.PixelArray
Vector2 = pygame.math.Vector2
Vector3 = pygame.math.Vector3
```

**Comment by Gaming32** [comment]
> hey folks, I'm running into some typing issues but I'm not sure if this is the right place to ask for help. I use `mypy` and it is complaining when I use `pygame.Surface` instead of `pygame.surface.Surface` as my types. Are these meant to be different types or is `pygame.Surface` supposed to be a shortcut to `pygame.surface.Surface`? I have the same issue with `pygame.Vector2` vs `pygame.math.Vector2`.
> 
> ```
> Incompatible types in assignment (expression has type "pygame.math.Vector2", variable has type "pygame.Vector2")
> Incompatible return value type (got "pygame.surface.Surface", expected "pygame.Surface")
> ```
> 
> I think the issue is because `pygame.Surface` seems to be defined as a class that extends `pygame.surface.Surface` instead of just being defined as the same class. So things like `pygame.display.set_mode()` will always return instances of `pygame.surface.Surface` and not `pygame.Surface`.
> 
> Can anyone help me better understand what I might be doing wrong? Are these meant to be different classes or is this a bug?

@Zeelot I am having the same issue. (maybe create an issue?)

**Comment by tusqasi** [comment]
The function here takes name which can also be `None`.
https://github.com/pygame/pygame/blob/main/src_py/sysfont.py#L403

But in the stub file it doesn't show it can also be `None`
https://github.com/pygame/pygame/blob/main/buildconfig/pygame-stubs/font.pyi#L18

When `name=None` pygame finds some fallback font or something but it works.

Should it made optional?

**Comment by tusqasi** [comment]
```py
name: Union[str, bytes, Iterable[Union[str, bytes]]]
```
to
```py
name: Union[str, bytes, Iterable[Union[str, bytes]], None]
```

**Comment by tusqasi** [comment]
> Measure load time. What is loadtime of a game with and without? (python2 vs python3.6/3.7/etc)  

I can try do this. But I have no clue how to proceed. :sweat_smile: .
Should I start a pygame project and measure the times,
and then delete the stub files in the site packages and now measure again. 

Let me know if you folk know any methods to do this

**Comment by MyreMylar** [comment]
Your testing seems reasonable to me. 

It is known that loading the typing module will increase the initial load of a script over not using it, but it looks like the performance penalty is very small in the context of the average game load time. It looks like in your test there it was about a maximum of tenth of second difference load time over ten runs? I suspect the average user is not going to notice if a game loads 1/100th of a second slower.

**Comment by adamsralf** [comment]
Hey, the line
 `  self.rect: pygame.Rect = self.image.get_rect()`
creates the pylance warning
  "pygame.rect.Rect" is incompatible with "pygame.Rect"
The line 
 `  self.rect: pygame.rect.Rect = self.image.get_rect()`
doesn't create a warning but is wrong, or not? 

BTW: In which source file is pygame.Rect defined?
Thanks, Ralf

**Comment by tusqasi** [comment]
https://github.com/pygame/pygame/tree/main/buildconfig%2Fstubs%2Fpygame

I think.

**Comment by ankith26** [comment]
This issue is fixed on main, IIRC... maybe even made it to the recent 2.1.2dev4 dev release, so maybe you can try upgrading to that first and see if the problem persists
